koku/api/tags/queries.py

```diff
@@ -195,71 +212,71 @@ def get_tag_keys(self, filters=True):
-        merged_data = []
-        for item in tag_keys:
-            for key, value in item.items():
-                key_dict = TagQueryHandler._get_dictionary_for_key(merged_data, key)
-                if not key_dict:
-                    new_dict = {}
-                    new_dict["key"] = key
-                    new_dict["values"] = [value]
-                    if source.get("type"):
-                        new_dict["type"] = source.get("type")
-                    merged_data.append(new_dict)
-                else:
-                    if value not in key_dict.get("values"):
-                        key_dict["values"].append(value)
-                        key_dict["values"].sort()
-        return merged_data
+                tag_keys.update({tag.get("key") for tag in tag_keys_query})
+
+        return list(tag_keys)
 
     def get_tags(self):
-        """Get a list of tags and values to validate filters."""
+        """Get a list of tags and values to validate filters.
+
+        Return a list of dictionaries containing the tag keys.
+        If OCP, these dicationaries will return as:
+            [
+                {"key": key1, "values": [value1, value2], "type": "storage" (or "pod")},
+                {"key": key2, "values": [value1, value2], "type": "storage" (or "pod")},
+                etc.
+            ]
+        If cloud provider, dicitonaries will be:
+            [
+                {"key": key1, "values": [value1, value2]},
+                {"key": key2, "values": [value1, value2]},
+                etc.
+            ]
+        """
         type_filter = self.parameters.get_filter("type")
 
-        merged_data = []
+        merged_data = defaultdict(list)
+        final_data = []
         with tenant_context(self.tenant):
-            tag_keys = []
+            tag_keys = {}
             for source in self.data_sources:
                 if type_filter and type_filter != source.get("type"):
                     continue
-                exclusion = self._get_exclusions(source.get("db_column"))
+                exclusion = self._get_exclusions("key")
                 tag_keys = (
                     source.get("db_table")
                     .objects.filter(self.query_filter)
                     .exclude(exclusion)
-                    .values(source.get("db_column"))
+                    .values("key", "values")
                     .distinct()
                     .all()
                 )
-                tag_keys = [tag.get(source.get("db_column")) for tag in tag_keys]
+                for dikt in tag_keys:
+                    merged_data[dikt.get("key")].extend(dikt.get("values"))
+
+                if source.get("type"):
+                    self.append_to_final_data_with_type(final_data, merged_data, source)
+                    merged_data = defaultdict(list)
+
+            if not source.get("type"):
+                self.append_to_final_data_without_type(final_data, merged_data)
```

Comment:
Need to think of better logic here. If we were to look at OCP and non-OCP tables at the same time, this logic would explode, and we would likely drop some tags somewhere.